Migrate 'Manage Groups' pages to Bootstrap

dmarcoux opened the pull request:
Links for review:
- https://obs-reviewlab.opensuse.org/dmarcoux-webui2-manage-groups/groups
- https://obs-reviewlab.opensuse.org/dmarcoux-webui2-manage-groups/group/new
- https://obs-reviewlab.opensuse.org/dmarcoux-webui2-manage-groups/group/edit/review-team
- https://obs-reviewlab.opensuse.org/dmarcoux-webui2-manage-groups/group/show/review-team

![group-index](https://user-images.githubusercontent.com/1102934/53729582-e1173100-3e75-11e9-8f4d-1eb95e1e7bd8.png)
![group-create](https://user-images.githubusercontent.com/1102934/53729580-e1173100-3e75-11e9-868d-e91a708897a6.png)
![group-edit](https://user-images.githubusercontent.com/1102934/53729581-e1173100-3e75-11e9-9abc-acd4a8fe234c.png)
![group-show](https://user-images.githubusercontent.com/1102934/53729583-e1afc780-3e75-11e9-926d-eae7f441bebb.png)

dmarcoux commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Manage Groups](https://trello.com/c/qkefDNxt/1052-manage-groups)

hennevogel commented:
CRUD! YAY! 🙃

bgeuken commented on src/api/app/views/webui2/webui/groups/new.html.haml:
```diff
@@ -0,0 +1,19 @@
+        = form.text_field(:title, class: 'form-control', required: true, autofocus: true)
+      .form-group
+        = label_tag(:members)
+        = text_field_tag('group[members]', @members, id: 'group_members', class: 'form-control tag-input', data: { source: user_autocomplete_path })
```
Use dashes in case of html ids and classes: `group-members`

dmarcoux replied:
Following your suggestion still renders the `input` with the ID `group_members`. I will revert this change since it's confusing to have the view with `group-members` and the JavaScript with `group_members` in when fetching the element with jQuery.

bgeuken replied:
> Following your suggestion still renders the `input` with the ID `group_members`.

Not on my machine. May it be that you removed the id completely? Because in that case it uses what `text_field_tag` generates which is 'group_members'. In other words you can drop it, if you stick with the underscore version.

> I will revert this change since it's confusing to have the view with `group-members` and the JavaScript with `group_members`

Well, you would have to update any place where the id is used:smiley:

I don't think there is any style guide that would deviate from using hyphens as delimiter for html classes and ids. Similar like ruby using snake_case and js camelCase.

dmarcoux replied:
I don't remember what I did last week, but it was clearly wrong. I guess I removed the `ID` and this is why it was still with the underscore. Anyway, I changed it now.

bgeuken commented on src/api/spec/bootstrap/features/webui/groups_spec.rb:
```diff
@@ -2,47 +2,74 @@
-    click_button('Accept')
-    expect(page).to have_content('eisendieter')
+    # Remove all users from tokenfield
+    2.times { first('button.tag-remove').click }
```
IIRC, you could use `find_all` instead

dmarcoux replied:
`find` does find all elements, but clicking them doesn't work because it says something like `Ambiguous match, found 2 elements matching (...)`.

bgeuken replied:
[all / find_all](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Finders#all-instance_method) yields the elements. So you probably could write `
all('button.tag-remove').each { |element| element.click }` (*untested*)

dmarcoux replied:
This worked, thank you!

dmarcoux commented:
@bgeuken Please have a look again, I commented on the suggestions you wrote.

bgeuken commented on src/api/app/assets/javascripts/webui2/tokenfield.js:
```diff
@@ -0,0 +1,381 @@
+/* jQuery Tags Input Revisited Plugin
```
This file should go to [vendor/assets](https://guides.rubyonrails.org/asset_pipeline.html#asset-organization)

dmarcoux replied:
I was wondering if we had something for this. Thank you for the pointer!

dmarcoux replied:
Thinking about this, I think this should also be done for CodeMirror (so {javascripts,stylesheets}/cm2  and collapse-text.scss files).

bgeuken commented on src/api/app/assets/stylesheets/webui2/tokenfield.scss:
```diff
@@ -0,0 +1,124 @@
+// Adapted from https://github.com/underovsky/jquery-tagsinput-revisited
```
This one as well.

coolo replied:
we really need a 'how to correctly use open source' training :(

bgeuken commented:
Great that you already ported the tests :+1:

Ana06 commented on src/api/app/views/webui2/webui/groups/edit.html.haml:
```diff
@@ -0,0 +1,15 @@
+:ruby
```
[nitpick] As it is just one line, you can use `-` instead :wink:

Ana06 commented on src/api/app/views/webui2/webui/groups/index.html.haml:
```diff
@@ -4,3 +4,28 @@
+        - @groups.each do |group|
+          %tr
+            %td
+              = link_to(group.title, group_show_path(title: group.title), id: group.title)
```
[nitpick] `.title` is not needed (and in the rest of the places where it is converted to string or param neither

Ana06 commented on src/api/app/views/webui2/webui/groups/index.html.haml:
```diff
@@ -4,3 +4,28 @@
+          %th Members
+          %th Action
+      %tbody
+        - @groups.each do |group|
```
N+1 queries, `includes(:users)`

dmarcoux replied:
It's already included in the controller

Ana06 replied:
This is because the `order` that is done in the database level. I wonder, do we really want to order by name? I would let it by when it was added, so that it doesn't change the order after saving. If we want to keep this ordering, I would suggest to remove the `includes` from the controller, which saves us 2 database queries (this should only be done in the new UI, as in the old one the order is not there, but just moving it to the old view would make it)

dmarcoux replied:
It's much better to order data alphabetically than by the creation date because it is constant. For example, the group 'Admins' will always be before 'OBS Developers'. Users would have to always use the search for finding a group because groups wouldn't be ordered. Nobody would remember after a few days when groups were created, so the order by creation date doesn't bring value at all.

dmarcoux replied:
I fixed the N+1 queries by adding `order(:login)` on the `users` association for the Group model. I don't see why we wouldn't want to order the users by login, so I believe it is fine. Let me know what you think.

Ana06 commented on src/api/app/views/webui2/webui/groups/index.html.haml:
```diff
@@ -4,3 +4,28 @@
+            %td
+              = link_to(group.title, group_show_path(title: group.title), id: group.title)
+            %td
+              = safe_join(group.users.order('users.login ASC').map { |user| link_to(truncate(user.login, length: 24), user_show_path(user)) }, ', ')
```
[nitpick] `ASC` is the default, so `order('users.login')` should be enough

Ana06 commented on src/api/app/views/webui2/shared/_group_members.html.haml:
```diff
@@ -0,0 +1,20 @@
+        %i.fas.fa-edit.text-secondary
+        Edit Group
+
+- if group.users.none?
```
[nitpick] using `blank?` instead saves as a query to the database as it is already cached for `group.users`

dmarcoux commented on src/api/app/assets/javascripts/webui2/tokenfield.js:
```diff
@@ -0,0 +1,381 @@
+ * https://github.com/underovsky/jquery-tagsinput-revisited
+ *
+ * Copyright (c) Krzysztof Rusnarczyk
+ * Licensed under the MIT license */
```
I will put it in the header of this file.

dmarcoux commented:
@bgeuken I adjusted the tokenfield. I still use the same color, which is `badge-secondary` from our patterns. I couldn't find something a lighter blue without it looking weird. I tried a few `alert-*` colors, but they all looked weird. The tokens are now based on the badges, so they should be pretty similar to what we use in other views. I also replaced the remove icon by the one we use normally to remove/delete. I would like to set the inside color of the icon to white, but I will leave it like this for now.

Let me know what you think.

bgeuken commented:
> @bgeuken I adjusted the tokenfield. I still use the same color, which is `badge-secondary` from our patterns.

secondary color sounds like the right thing to use. I guess it's just that the font size is much smaller than for badges.

I had a brief look; badge texts have `font-weight: 700;`.

With font-weight: inherited (aka. currently)
![2019-03-01_17-55](https://user-images.githubusercontent.com/968949/53653270-3fac9700-3c4b-11e9-9e2d-333e1fca8160.png)

With font-weight: 700 (aka bootstrap badges):
![2019-03-01_17-55_1](https://user-images.githubusercontent.com/968949/53653276-43d8b480-3c4b-11e9-9c67-c606ecb9f686.png)

I think it looks nicer with the changed font-weight and it's worth changing it. But I am not blocking the PR on this.

bgeuken commented:
The other thing I noticed is that the text of group members is quite smaller (like [here](https://obs-reviewlab.opensuse.org/dmarcoux-webui2-manage-groups/group/show/review-team)). IMO it should have the same font size we use elsewhere.

![2019-03-01_18-01](https://user-images.githubusercontent.com/968949/53653618-00cb1100-3c4c-11e9-9984-ac980bfef285.png)

bgeuken commented:
And on wider screens the text fields get very wide:
![2019-03-01_18-04](https://user-images.githubusercontent.com/968949/53653815-8484fd80-3c4c-11e9-967d-bbb044771dcd.png)

On the other views we limited the size by adding a col-md-6.

dmarcoux commented:
Thank you @bgeuken and @Ana06 for the feedback. I'm going to address your requests today.

dmarcoux commented:
@Ana06 and @bgeuken, please have another look.

bgeuken commented:
@dmarcoux One minor thing I noticed is that the group edit view now has the configuration tabs. That's different from the user edit view. We should align this one way or the other.

dmarcoux commented:
> @dmarcoux One minor thing I noticed is that the group edit view now has the configuration tabs. That's different from the user edit view. We should align this one way or the other.

The `Manage Groups` tab should be active for the group new/edit views, right? Is this what you mean?

Ana06 commented on src/api/app/models/group.rb:
```diff
@@ -7,7 +7,7 @@
   has_one :staging_workflow, class_name: 'Staging::Workflow', foreign_key: :managers_group_id, dependent: :nullify
   has_many :groups_users, inverse_of: :group, dependent: :destroy
-  has_many :users, -> { distinct }, through: :groups_users
+  has_many :users, -> { distinct.order(:login) }, through: :groups_users
```
nice solution! :+1: It is also affecting the old UI, but I think it is ok

Ana06 replied:
btw, `includes(:users)` (instead of `includes(:groups_users, :users)`) in the controller is enough. (although not necessarily part of this PR)

dmarcoux replied:
Good catch, I will fix it.

dmarcoux commented:
I need to fix errors in OBS Package Build.

hennevogel commented:
> I need to fix errors in OBS Package Build.

No you don't. This is not a required check.

dmarcoux commented:
> No you don't. This is not a required check.

I believe it's already fixed. It's just a matter of listing the directory under the packaged files.

bgeuken commented:
> dmarcoux commented 2 hours ago
> > @dmarcoux One minor thing I noticed is that the group edit view now has the configuration tabs. That's different from the user edit view. We should align this one way or the other.
>
> The Manage Groups tab should be active for the group new/edit views, right? Is this what you mean?

Sorry, I've overlooked this. What I meant is that the user edit has no configuration tab:

![2019-03-04_14-15](https://user-images.githubusercontent.com/968949/53735764-204e7d80-3e88-11e9-83f3-9015b6c1654e.png)

But the group edit does:

![2019-03-04_14-14](https://user-images.githubusercontent.com/968949/53735773-29d7e580-3e88-11e9-8b16-9cbdf5674f13.png)

I would align the two. I have a small preference to not show the tab, because edit and create (for users) are basically sub-tabs. They are not directly connected to the configuration tab.

That said, I think this is something we can decide in a separate PR.

dmarcoux commented:
@bgeuken It's clearer now. I also prefer without tabs as the breadcrumbs are more prominent and tabs don't really bring value in this context. I'll do it in this PR.
